adaptdl/adaptdl/torch/scaling_rules.py

```diff
@@ -91,9 +76,84 @@ def __init__(self, adp, optimizer, num_replicas=None,
         self._num_replicas = (num_replicas if num_replicas is not None
                               else torch.distributed.get_world_size())
         self._accum_scale = accum_scale or self._num_replicas
+
+        if patch_optimizer:
+            self.patch_optimizer()
+
+        self._reset_accumulation()
+
+    def _reset_accumulation(self):
+        self._orig_optimizer_zero_grad()
+        self._local_sqr = None
+        self._accum_count = 0
+
+    @property
+    def accum_scale(self):
+        return self._accum_scale
+
+    def set_accum_scale(self, accum_scale):
+        if not np.isclose(self._accum_scale, accum_scale):
+            self._reset_accumulation()
+            self._accum_scale = accum_scale
+
+    def _calculate_lr_factors(self, scale):
+        raise NotImplementedError
+
+    def calculate_lr_factor_estimation(self, scale):
+        raise NotImplementedError
+
+    def step(self, *args, **kwargs):
+        """
+        Run one optimizer step. Essentially just invokes
+        ``optimizer.step(*args, **kwargs)`` with a scaled learning rate.
+
+        Arguments:
+            args: Positional arguments passed to ``optimizer.step``.
+            kwargs: Keyword arguments passed to ``optimizer.step``.
+        """
+        raise NotImplementedError
+
+    def zero_grad(self, *args, **kwargs):
+        if self._should_zero_grad:
+            self._orig_optimizer_zero_grad(*args, **kwargs)
+        else:
+            warnings.warn("skipping zero_grad for accumulated gradient")
+
+    def patch_optimizer(self):
+        """
+        Monkey-patch the optimizer's step function with
+        :meth:`_ScalingRuleBase.step`.
+        """
+        # TODO: detect if the optimizer has already been patched.
+
+        @functools.wraps(self._optimizer.step)
+        def step_wrapper(optim, *args, **kwargs):
+            return self.step(*args, **kwargs)
+
+        @functools.wraps(self._optimizer.zero_grad)
+        def zero_wrapper(optim, *args, **kwargs):
+            return self.zero_grad(*args, **kwargs)
+        self._optimizer.step = MethodType(step_wrapper, self._optimizer)
+        self._optimizer.zero_grad = MethodType(zero_wrapper, self._optimizer)
+
+    def to_tensorboard(self, writer, global_step, tag_prefix):
+        """Output some useful metrics to TensorBoard."""
+        pass
+
+
+_Base = _ScalingRuleBase if typing.TYPE_CHECKING else object
+
+
+class _GradientNoiseScaleMixin(_Base):
+    """This mixin class makes _ScalingRuleBase able to track gradient related
+    stats."""
```

Comment:
Not sure this class makes sense to be a Mixin. Alternative is an object owned by AdaptiveDataParallel.

Reply:
It is challenging to make it a stand-alone class owned by AdaptiveDataParallel mainly because GradientNoiseScale and ScalingRule depend on each other and is forming a cyclic dependency.
E.g. GradientNoiseScale needs ScalingRuleBase._accum_count https://github.com/petuum/adaptdl/blob/9b4ef6417038703c07d2c85b405935e0eb6a609c/adaptdl/adaptdl/torch/scaling_rules.py#L284
and ScalingRuleBase needs GradientNoiseScale.sqr_avg.
https://github.com/petuum/adaptdl/blob/9b4ef6417038703c07d2c85b405935e0eb6a609c/adaptdl/adaptdl/torch/scaling_rules.py#L211
I.e, we'll need to awkwardly create the two objects in a way like:
```python
gns = GradientNoiseScale(...)
sr = ScalingRule(...)
gns.sr = sr
sr.gns = gns
```
Unless we make accumulation steps related code a part of GradientNoiseScale, I don't see an easy way to completely remove the cyclic dependency.

PS. I simply moved all the callback functions to GradientNoiseScale which is not ideal because some accumulation steps related code is moved there altogether. But even if I further factor them out, the two examples above will remain there and cause the problem.

Reply:
Actually the gradient accumulation code _should_ be primarily in the gradient noise scale class. The lr scaler shouldn't need to know about gradient accumulation much at all.

The code snippet you posted looks fine, and can be hidden away in the AdaptiveDataParallel's `__init__` method, i.e.
```
self.scaling_rule = scaling_rule
self.scaling_rule.adp = self
```